File: site/wsgi/load.py
#!/usr/bin/env python
# load.py

import sys, os, cgi, re, json, stat, time
sys.path.append(os.path.dirname(__file__))
import config

rootcachename = os.path.join(config.cachedir, 'rootcache')

def uri_without_query(uri):
  return uri.split('?', 1)[0]

def splituri(uri):
  return [p for p in uri_without_query(uri).split('/') if p]

def filename_from_uri(uri):
  return '/'.join(splituri(uri)[1:])

def format_from_uri(uri):
  if 'load' == splituri(uri)[0]:
    return 'json'
  elif 'download' == splituri(uri)[0]:
    return 'download'
  return ''
  

def application(env, start_response):
  form = cgi.FieldStorage(fp=env['wsgi.input'], environ=env)
  request_uri = env['REQUEST_URI']
  host_name = env['HTTP_HOST']
  filename = form.getfirst("file", filename_from_uri(request_uri))
  format = form.getfirst("format", format_from_uri(request_uri))
  callback = form.getfirst("callback", None)
  tail = form.getfirst("tail", None)
  user = '.'.join(host_name.split('.')[:-2])
  domain = '.'.join(host_name.split('.')[-2:])
  origfilename = filename
  if len(user) is 0:
    user = None
  
  try:
    tail = int(tail)
  except:
    tail = None
  
  isrootlisting = (user is None and filename == "" and format == "json")
  if isrootlisting:
    try:
      data = file(rootcachename, 'rb').read()
      start_response('200 OK', [
         ('Cache-Control', 'no-cache, must-revalidate'),
         ('Content-Type', 'text/javascript')
      ])
      if callback is not None:
        return [callback, '(', data, ')']
      else:
        return [data]
    except:
     pass
  
  def jsonout(dict):
    dumps = json.dumps(dict)
    if isrootlisting:
      cache = file(rootcachename, 'wb')
      cache.write(dumps)
      cache.close()
    if callback is not None:
      return [callback, '(', dumps, ')']
    else:
      return [dumps]
  
  def errorexit(s):
    if format == 'json':
      start_response('200 OK', [
         ('Cache-Control', 'no-cache, must-revalidate'),
         ('Content-Type', 'text/javascript')
      ])
      return jsonout({'error': s})
    else:
      start_response('200 OK', [
         ('Content-Type', 'text/html')
      ])
      return ['<plaintext>', s]
  
  def mimetype(ext, filename, data=None):
    result = {
      'jpg'  : 'image/jpeg',
      'jpeg' : 'image/jpeg',
      'gif'  : 'image/gif',
      'png'  : 'image/png',
      'bmp'  : 'image/x-ms-bmp',
      'ico'  : 'image/x-icon',
      'htm'  : 'text/html',
      'html' : 'text/html',
      'txt'  : 'text/plain',
      'text' : 'text/plain',
      'css'  : 'text/css',
      'coffee' : 'text/coffeescript',
      'js'   : 'text/javascript',
      'xml'  : 'text/xml'
    }.get(ext, None)
    if result is None:
      result = 'text/x-turtlebits'
    if result.startswith('text'):
      result += ';charset=utf-8'
    return result
  
  # Validate a good username
  if user is not None:
    if not re.match(r'^[A-Za-z]\w{2,}$', user):
      return errorexit('bad username "%s"' % user)
    filename = os.path.join(user, filename)
  # Validate a good filename
  if (len(filename) and
      not re.match(r'^(?:[\w][\w\.\-]*)(?:/[\w][\w\.\-]*)*/?$', filename)):
    return errorexit('bad filename "%s"' % filename)
  if (len(filename) and
      not re.match(r'^(?:[\w][\w\.\-]*)(?:/[\w][\w\.\-]*)*/?$', filename)):
    return errorexit('bad filename "%s"' % filename)
  absfile = os.path.join(config.datadir, filename)
  if not absfile.startswith(config.datadir):
    return errorexit('illegal filename "%s"' % filename)
  
  def validnewfile(abs):
    while True:
      if os.path.isfile(abs):
        return False
      abs = os.path.dirname(abs)
      if not absfile.startswith(config.datadir):
        return False
      if os.path.isdir(abs):
        return True
  
  def direntry(absdir, filename):
    fn = os.path.join(absdir, filename)
    fs = os.stat(fn)
    (mode, ino, dev, nlink, uid, gid, size, atime, mtime, ctime) = fs
    modestr = ''
    if stat.S_ISDIR(mode): modestr += 'd'
    if mode & stat.S_IRUSR: modestr += 'r'
    if mode & stat.S_IWUSR: modestr += 'w'
    if mode & stat.S_IXUSR: modestr += 'x'
    mtime = fs.st_mtime
    if 'd' not in modestr and size == 0:
      mtime = 0
    return { 'name': filename, 'mode': modestr,
      'size': fs.st_size, 'mtime': mtime }
  
  def readtail(f, lines):
    if lines <= 0:
      return ''
    f.seek(0, 2)
    bytes = f.tell()
    size = lines
    block = -1
    while size > 0:
      truncate = min(0, bytes + block * 1024)
      f.seek(block * 1024 - truncate, 2) # from the end
      data = f.read(truncate + 1024)
      linesFound = data.count('\n')
      size -= linesFound
      block -= 1
      if truncate: break
    truncate = min(0, bytes + block * 1024)
    f.seek(block * 1024 - truncate, 2)
    while size < 0:
      f.readline()
      size += 1
    return f.read()
  
  def userhaskey(user):
    if not user:
      return False
    keydir = os.path.join(config.datadir, user, '.key')
    if not os.path.isdir(keydir):
      return False
    keys = os.listdir(keydir)
    return len(keys) > 0
  
  if format == 'json':
    haskey = userhaskey(user)
    if os.path.isfile(absfile):
      f = file(absfile, 'r')
      if tail is not None:
        data = readtail(f, tail)
      else:
        data = f.read()
      fs = os.fstat(f.fileno())
      (mode, ino, dev, nlink, uid, gid, size, atime, mtime, ctime) = (
          os.fstat(f.fileno()))
      f.close()
      ext = filename.rsplit('.', 1)[-1]
      mimet = mimetype(ext, absfile, data)
      start_response('200 OK', [
         ('Cache-Control', 'no-cache, must-revalidate'),
         ('Content-Type', 'text/javascript')
      ])
      return jsonout({'file': '/' + filename, 'data': data,
          'auth': haskey, 'mtime': fs.st_mtime, 'mime': mimet})
    if os.path.isdir(absfile):
      if len(filename) and not filename.endswith('/'):
        filename += '/'
      contents = os.listdir(absfile) or []
      contents = sorted([fn for fn in contents if not fn.startswith('.')])
      list = [direntry(absfile, n) for n in contents]
      start_response('200 OK', [
         ('Cache-Control', 'no-cache, must-revalidate'),
         ('Content-Type', 'text/javascript')
      ])
      return jsonout({'directory': '/' + filename,
                     'list': list,
                     'auth': haskey})
    if len(filename) and not filename.endswith('/') and validnewfile(absfile):
      start_response('200 OK', [
         ('Cache-Control', 'no-cache, must-revalidate'),
         ('Content-Type', 'text/javascript')
      ])
      return jsonout({'error': 'could not read file %s' % filename,
                     'newfile': True,
                     'auth': haskey,
                     'info': absfile})
    return errorexit('could not read file %s' % filename)
  
  if format == 'download' and os.path.isfile(absfile):
    shortfilename = re.sub('^.*/', '', filename)
    if tail is not None:
      data = readtail(file(absfile, 'r'), tail)
    else:
      data = file(absfile, 'r').read()
    start_response('200 OK', [
       ('Cache-Control', 'no-cache, must-revalidate'),
       ('Content-Type', 'application/octet-stream'),
       ('Content-Disposition', 'attachment; filename=%s' % shortfilename),
       ('Content-Length', '%d' % len(data))
    ])
    return data
  
  def wrapturtle(text):
    return (
    '<!doctype html>\n<html>\n<head>\n' +
    '<script src="http://%s/turtlebits.js"></script>\n' % (domain) +
    '</head>\n<body><script type="text/coffeescript">\neval $.turtle()\n\n' +
    text + '\n</script></body></html>')
  
  if os.path.isfile(absfile):
    split = filename.rsplit('.', 1)
    ext = None if len(split) < 2 else split[-1]
    data = file(absfile, 'r').read()
    mt = mimetype(ext, absfile, data)
    # for turtlebits: if there is no extension and it looks like a non-HTML
    # text file, assume it is coffeescript that should be wrapped in HTML.
    if mt.startswith('text/x-turtlebits'):
      data = wrapturtle(data)
      mt = mt.replace('x-turtlebits', 'html')
    start_response('200 OK', [
       ('Cache-Control', 'no-cache'),
       ('Content-Type', mt)
    ])
    return data
  
  elif os.path.isdir(absfile) or not '/' in filename.rstrip('/'):
    if len(filename) and not filename.endswith('/'):
      start_response('301 Redirect', [('Location', request_uri + '/')])
      return []
    start_response('200 OK', [
       ('Cache-Control', 'no-cache'),
       ('Content-Type', 'text/html;charset=utf-8')
    ])
    result = []
    result.append('<title>%s</title>\n' % (request_uri))
    result.append('<style>\n')
    result.append('body { font-family:Arial,sans-serif; }\n')
    result.append('pre {\n')
    result.append('-moz-column-count:3;\n')
    result.append('-webkit-column-count:3;\n')
    result.append('column-count:3;\n')
    result.append('}\n')
    result.append('</style>\n')
    result.append(absfile)
    result.append('<h3>%s</h3>\n' % (env['HTTP_HOST'] + request_uri))
  
    contents = os.path.isdir(absfile) and os.listdir(absfile) or []
    contents = [fn for fn in contents if not fn.startswith('.')]
    result.append('<pre>\n')
    if len(request_uri.strip('/').split('/')) > 1:
      result.append('<a href="%s" style="background:yellow">Up to %s</a>\n' % (
        os.path.dirname(request_uri.rstrip('/')) + '/',
        os.path.dirname(request_uri.rstrip('/')) + '/'))
    for name in sorted(contents):
      af = os.path.join(absfile, name)
      if os.path.isdir(af) and not name.endswith('/'):
        name += '/'
      link = '<a href="%s">%s</a>' % (request_uri + name, name)
      if os.path.isfile(af):
        link += (' <a href="%s" style="color:lightgray" rel="nofollow">edit</a>'
                % ('/edit/' + filename + name))
      result.append(link + '\n')
    result.append('</pre>\n')
  
    if len(contents) == 0:
      result.append('(directory is empty)<br>\n')
    return result
  elif filename.endswith('/'):
    start_response('301 Redirect', [('Location',
        os.path.dirname(request_uri.rstrip('/')) + '/')])
    return []
  
  else:
    start_response('200 OK', [
       ('Cache-Control', 'no-cache'),
       ('Content-Type', 'text/html;charset=utf-8')
    ])
    result = []
    result.append('<pre>\n')
    result.append('No file %s found.\n\n' % origfilename)
    result.append('<a href="%s">Up to %s</a>\n' % (
        os.path.dirname(request_uri.rstrip('/')) + '/',
        os.path.dirname(request_uri.rstrip('/')) + '/'))
    if filename.rsplit('.', 1)[-1] in ['htm', 'html', 'js', 'css']:
      print
      result.append('<a href="%s" rel="nofollow">Create %s</a>\n' % (
          '/edit/' + origfilename, '/home/' + origfilename))
    result.append('</pre>\n')
    return result

File: dev/load.js
var path = require('path');
var fs = require('fs-ext');
var fsExtra = require('fs.extra');
var utils = require('./utils');

exports.handleLoad = function(req, res, app, format) {
  var filename = req.param('file', utils.filenameFromUri(req));
  var callback = req.param('callback', null);
  var tail = req.param('tail', null);
  var user = utils.getUser(req, app);
  var origfilename = filename;

  tail = parseInt(tail);
  if (Number.isNaN(tail)) {
    tail = null;
  }

  if (filename == null) {
    filename = '';
  }

  try {
    var isrootlisting = !user && filename == '' && format == 'json';

    //console.log({'filename': filename, 'user': user, 'isroot': isrootlisting});

    if (isrootlisting) {
      try {
        // Try the cache first if it exists
        var data = fsExtra.readJsonSync(utils.getRootCacheName(app));

        res.set('Cache-Control', 'no-cache, must-revalidate');
        res.set('Content-Type', 'text/javascript');

        //console.log(data);
        res.jsonp(data);
        return;
      }
      catch (e) { }
    }

    // Validate username
    if (user) {
      utils.validateUserName(user);
      if (filename.length > 0) {
        filename = path.join(user, filename);
      }
      else {
        filename = user + '/';
      }
    }

    // Validate filename
    if (filename.length > 0) { 
        if (!utils.isFileNameValid(filename, false)) {
        utils.errorExit('Bad filename: ' + filename);
      }
    }

    var absfile = utils.makeAbsolute(filename, app);

    if (format == 'json') {
      var haskey = userhaskey(user, app);

      // Handle the case of a file that's present
      if (utils.isPresent(absfile, 'file')) {
        var data = (tail != null) ? 
            readtail(absfile, tail) : fs.readFileSync(absfile, {'encoding': 'utf8'});

        var statObj = fs.statSync(absfile);

        var mimetype = 
            getMimeType(filename.substring(filename.lastIndexOf('.')));

        res.set('Cache-Control', 'no-cache, must-revalidate');
        res.jsonp({'file': '/' + filename, 
                   'data': data,
                   'auth': haskey,
                   'mtime': statObj.mtime.getTime(), 
                   'mime': mimetype});
        return;
      }
    
      // Handle the case of a directory that's present
      if (utils.isPresent(absfile, 'dir')) {
        if (filename.length > 0 && filename[filename.length - 1] != '/') {
          filename += '/';
        }

        var list = buildDirList(absfile, fs.readdirSync(absfile).sort());

        var jsonRet = 
          {'directory': '/' + filename, 'list': list, 'auth': haskey};

        // Write to cache if this is a top dir listing
        if (isrootlisting) {
          fsExtra.outputJsonSync(utils.getRootCacheName(app), jsonRet);
        }
        res.set('Cache-Control', 'no-cache, must-revalidate');
        res.jsonp(jsonRet);
        return;
      }

      // Handle the case of a new file create
      if (filename.length > 0 && 
          filename[filename.length - 1] != '/' && 
          isValidNewFile(absfile, app)) {
        res.set('Cache-Control', 'no-cache, must-revalidate');
        res.jsonp({'error': 'could not read file ' + filename,
                   'newfile': true,
                   'auth': haskey,
                   'info': absfile});
        return;
      }

      utils.errorExit('Could not read file ' + filename);
    }
    else if (format == 'execute') { // File loading outside the editor
      if (utils.isPresent(absfile, 'file')) {
        var mt = 
            getMimeType(filename.substring(filename.lastIndexOf('.')));
        var data = fs.readFileSync(absfile, {'encoding': 'utf8'});

        //For turtle bits, assume it's coffeescript
        if (mt.indexOf('text/x-turtlebits') == 0) {
          data = wrapTurtle(data, app);
          mt = mt.replace('x-turtlebits', 'html');
        }

        res.set('Cache-Control', 'no-cache');
        res.set('Content-Type', mt);
        res.send(data);
        return;
      }
      else if (utils.isPresent(absfile, 'dir') || 
               filename.indexOf('/') == filename.length) {
        if (filename.length > 0 && filename[filename.length - 1] != '/') {
          res.redirect(301, '/home' + req.path + '/');
          return;
        }

        res.set('Cache-Control', 'no-cache');
        res.set('Content-Type', 'text/html;charset=utf-8');

        var text = '<title>' + req.path + '</title>';
        text += '<style>\n';
        text += 'body { font-family:Arial,sans-serif; }\n';
        text += 'pre {\n';
        text += '-moz-column-count:3;\n';
        text += '-webkit-column-count:3;\n';
        text += 'column-count:3;\n';
        text += '}\n';
        text += '</style>\n';
        text += '<h3>' + req.host + req.path + '</h3>\n';
        text += '<pre>\n';

        var lastIndex = req.path.lastIndexOf('/');
        if (lastIndex > 0) {
          var prevIndex = req.path.substring(0, lastIndex).lastIndexOf('/');
          if (prevIndex >= 0) {
            var parentDir = req.path.substring(0, prevIndex + 1);
            text += '<a href="/home' + parentDir;
            text += '" style="background:yellow">Up to ' + parentDir;
            text += '</a>\n';
          }
        }
        
        var contents = (utils.isPresent(absfile, 'dir')) ? 
          fs.readdirSync(absfile).sort() : new Array();
        for (var i = 0; i < contents.length; i++) {
          if (contents[i][0] == '.') {
            // Skip past any dirs starting with a '.'
            continue;
          }
          
          var name = contents[i];
          var af = path.join(absfile, name);
          if (utils.isPresent(af, 'dir') && 
              name.charAt(name.length - 1) != '/') {
            name += '/';
          }
          var link = '<a href="/home' + req.path + name + '">' + name + '</a>';
          if (utils.isPresent(af, 'file')) {
            link += ' <a href="/edit' + req.path + name + '" style="color:lightgray" rel="nofollow">edit</a>';
          }
          text += link + '\n';
        }
        text += '</pre>\n';
        if (contents.length == 0) {
          text += '(directory is empty)<br>\n';
        }

        res.send(text);
        return;
      }
      else if (filename.charAt(filename.length - 1) == '/') {
        res.redirect(301, 
                     path.dirname(filename.substring(0, filename.length-1)) + '/');
        return;
      }
      else {
        res.set('Cache-Control', 'no-cache');
        res.set('Content-Type', 'text/html;charset=utf-8');

        var text = '<pre>\n';
        text += 'No file ' + origfilename + ' found.\n\n';
        var strip = (req.path.charAt(req.path.length - 1) == '/') ?
          req.path.substring(0, req.path.length - 1) : req.path;
        var parentDir = path.dirname(strip) + '/';
        text += '<a href="' + parentDir + '">Up to ' + parentDir + '</a>\n';
        
        var extIdx = filename.lastIndexOf('.');
        if (extIdx > 0) {
          var ext = filename.substring(extIdx + 1);
          if (ext == 'htm' || ext == 'html' || ext == 'js' || ext == 'css') {
            text += '<a href="/edit/' + origfilename + '"> rel="nofollow">Create /home/' + origfilename + '</a>\n';
          }
        }
        text += '</pre>\n';
        res.send(text);
        return;
      }
    }
  }
  catch (e) {
    if (e instanceof utils.ImmediateReturnError) {
      if (format == 'json') {
        res.jsonp((e.jsonObj) ? e.jsonObj : {'error': e.toString()});
      }
      else {
        res.set('Content-Type', 'text/html');
        res.send('<html><body><plaintext>' + e.message);
      }
    }
    else {
      throw e;
    }
  }

};

function wrapTurtle(text, app) {
  return (
    '<!doctype html>\n<html>\n<head>\n' +
    '<script src="http://' + app.locals.config.host + 
    '/turtlebits.js"></script>\n' +
    '</head>\n<body><script type="text/coffeescript">\neval $.turtle()\n\n' +
    text + '\n</script></body></html>');
}

function isValidNewFile(newAbsFileName, app) {
  var dir = path.dirname(newAbsFileName);
  while (true) {
    if (dir.indexOf(app.locals.config.dirs.datadir) != 0) {
      return false;
    }

    try {
      if (fs.statSync(dir).isDirectory()) {
        return true;
      }
    }
    catch (e) { }
    dir = path.dirname(dir);
  }
}

function buildDirList(absdir, contents) {
  var list = new Array();

  for (var i = 0; i < contents.length; i++) {
    // Skip over any entries starting with . 
    if (contents[i][0] == '.') {
      continue;
    }

    var item = path.join(absdir, contents[i]);
    var statObj = fs.statSync(item);
    var modestr = '';
    if (statObj.isDirectory()) {
      modestr += 'd';
    }
    if (statObj.mode & 00400) {
      modestr += 'r';
    }
    if (statObj.mode & 00200) {
      modestr += 'w';
    }
    if (statObj.mode & 00100) {
      modestr += 'x';
    }

    var mtime = statObj.mtime.getTime();
    // If its an empty file, then reset mtime
    if (statObj.isFile() && statObj.size == 0) {
      mtime = 0;
    }

    list.push({'name': contents[i], 
               'mode': modestr, 
               'size': statObj.size,
               'mtime': mtime});
  }

  return list;
}

function readtail(filename, lines) {
  if (lines <= 0) {
    return '';
  }

  var fd = fs.openSync(filename, 'r');
  // seek to the end
  fd.seekSync(fd, 0, 2);

  // TODO : Implement this
}

function userhaskey(user, app) {
  if (!user) {
    return false;
  }

  var keydir = utils.getKeyDir(user, app);

  if (!utils.isPresent(keydir, 'dir')) {
    return false;
  }

  var keys = fs.readdirSync(keydir);
  if (!keys || keys.length == 0) {
    return false;
  }

  return true;
}

function getMimeType(ext) {
  var mimeTypeTable = { 'jpg'  : 'image/jpeg',
                        'jpeg' : 'image/jpeg',
                        'gif'  : 'image/gif',
                        'png'  : 'image/png',
                        'bmp'  : 'image/x-ms-bmp',
                        'ico'  : 'image/x-icon',
                        'htm'  : 'text/html',
                        'html' : 'text/html',
                        'txt'  : 'text/plain',
                        'text' : 'text/plain',
                        'css'  : 'text/css',
                        'coffee' : 'text/coffeescript',
                        'js'   : 'text/javascript',
                        'xml'  : 'text/xml' };
  var result = mimeTypeTable[ext];
  if (!result) {
    result = 'text/x-turtlebits';
  }

  if (result.indexOf('text') == 0) {
    result = result + ';charset=utf-8';
  }

  return result;
}